src/libaktualizr/package_manager/androidmanager.cc

```diff
@@ -0,0 +1,102 @@
+  package["version"] = "n/a";
+  packages.append(package);
+  return packages;
+}
+
+Uptane::Target AndroidManager::getCurrent() const {
+  std::vector<Uptane::Target> installed_versions;
+  std::string current_hash;
+  size_t current_k = SIZE_MAX;
+  storage_->loadPrimaryInstalledVersions(&installed_versions, &current_k, nullptr);
+  if (current_k != SIZE_MAX) {
+    current_hash = installed_versions[current_k].sha256Hash();
+  }
+  AndroidInstallationState installation_state;
+  if (AndroidInstallationDispatcher::GetState().state_ == AndroidInstallationState::STATE_INSTALLED) {
+    installation_state = AndroidInstallationDispatcher::Dispatch();
+    if (installation_state.state_ == AndroidInstallationState::STATE_UPDATE) {
+      current_hash = installation_state.payload_;
+    }
+  }
+  std::vector<Uptane::Target>::iterator it;
+  for (it = installed_versions.begin(); it != installed_versions.end(); it++) {
+    if (it->sha256Hash() == current_hash) {
+      if (installation_state.state_ == AndroidInstallationState::STATE_UPDATE) {
+        storage_->savePrimaryInstalledVersion(*it, InstalledVersionUpdateMode::kCurrent);
+        auto outcome = data::InstallOutcome(data::UpdateResultCode::kOk, "package successfully intalled");
+        storage_->storeInstallationResult(data::OperationResult(it->filename(), outcome));
+      }
+      return *it;
+    }
+  }
+  return getUnknown();
+}
+
+data::InstallOutcome AndroidManager::install(const Uptane::Target& target) const {
+  if (AndroidInstallationDispatcher::GetState().state_ == AndroidInstallationState::STATE_NOP) {
+    LOG_INFO << "Begin Android package installation";
+    storage_->savePrimaryInstalledVersion(target, InstalledVersionUpdateMode::kPending);
+    auto package_filename = (Utils::getStorageRootPath() / target.filename()).string() + "." + target.sha256Hash();
+    std::ofstream package_file(package_filename.c_str());
+    if (!package_file.good()) {
+      throw std::runtime_error(std::string("Error opening file ") + package_filename);
+    }
+    package_file << *storage_->openTargetFile(target);
+    AndroidInstallationDispatcher::Dispatch(package_filename);
+  }
+  return data::InstallOutcome(data::UpdateResultCode::kInProgress, "Installation is in progress");
+}
+
+AndroidInstallationState AndroidInstallationDispatcher::GetState() {
+  boost::filesystem::directory_iterator entryItEnd, entryIt(Utils::getStorageRootPath());
+  for (; entryIt != entryItEnd; ++entryIt) {
+    auto& entry_path = entryIt->path();
+    if (boost::filesystem::is_directory(*entryIt)) {
+      continue;
+    }
+    auto package_file_path = entry_path.string();
+    auto ext = entry_path.extension().string();
+    if (ext == ".installed") {
+      return {AndroidInstallationState::STATE_INSTALLED, package_file_path};
+    } else if (ext == ".inprogress") {
+      return {AndroidInstallationState::STATE_IN_PROGRESS, package_file_path};
+    } else if (true /* check ext is actually a hash */) {
```

Comment:
Is this still a TODO? What's the challenge there?

Reply:
No challenge, I'll restore the implementation soon.